source/module_basis/module_nao/hydrogen_radials.cpp

```diff
@@ -0,0 +1,412 @@
+    int istep = 1;
+    double delta_norm = 1.0;
+    while((std::fabs(delta_norm) > conv_thr))
+    {
+        rmin_ = rmax_;
+        rmax_ += delta_r;
+        int ngrid = static_cast<int>((rmax_ - rmin_) / dr) + 1; // [rmin, rmax]
+        std::vector<double> rgrid_segment(ngrid);
+        for(int ir = 0; ir != ngrid; ++ir)
+        {
+            rgrid_segment[ir] = rmin_ + ir * dr;
+        }
+        std::vector<double> rvalue_segment = generate_hydrogen_radial_segment(
+            charge, n, l, rmin_, rmax_, dr, rank, ptr_log);
+        // before push back, pop back the last element
+        if(rgrid.size() != 0)
+        {
+            rgrid.pop_back();
+            rvalue.pop_back();
+        }
+        rgrid.insert(rgrid.end(), rgrid_segment.begin(), rgrid_segment.end());
+        rvalue.insert(rvalue.end(), rvalue_segment.begin(), rvalue_segment.end());
+        delta_norm = norm;
+        norm = radial_norm(rgrid, rvalue);
+        delta_norm = norm - delta_norm;
+        if(istep == 0) printf("%10d%12.2f%14.10f%18.10e\n", istep, rmax_, norm, delta_norm);
+        ++istep;
+    }
+    printf("%10d%12.2f%14.10f%18.10e\n", istep, rmax_, norm, delta_norm);
+    return rmax_;
+}
+
+std::vector<std::pair<int, int>> HydrogenRadials::unzip_strategy(const int nmax,
+                                                                 const std::string strategy)
+{
+    if(strategy != "energy")
+    {
+        assert(nmax < 8);
+    }
+    std::vector<std::pair<int, int>> nl_pairs;
+    if(strategy == "minimal")
+    {
+        for(int n = 1; n <= nmax; n++)
+        {
+            std::pair<int, int> nl_pair = std::make_pair(n, n - 1);
+            nl_pairs.push_back(nl_pair);
+        }
+    }
+    else if(strategy == "energy")
+    {
+        // 1s, -(n+1)-> 2s, -(l+1)-> 2p, 3s, -(n+1)-> 3p, 4s, -(l+1)-> 3d, 4p, 5s, -(n+1)-> 4d, 5p, 6s
+        int starting_n = 1;
+        int starting_l = 0;
+        int nelec = 0;
+        int nl_switch = 0;
+        while(nelec < nmax) // now nmax is the full number of electrons of certain atom
+        {
+            int n = starting_n;
+            int l = starting_l;
+            do
+            {
+                nl_pairs.push_back(std::make_pair(n, l));
+                nelec += (2*l + 1)*2;
+                n++;
+                l--;
+            } while ((l >= 0)&&(nelec < nmax));
+            
+            if(nl_switch%2 == 1)
+            {
+                starting_l++;
+            }
+            else
+            {
+                starting_n++;
+            }
+            nl_switch++;
+        }
+    }
+    else
+    {
+        for(int n = 1; n <= nmax; n++)
+        {
+            for(int l = 0; l < n; l++)
+            {
+                std::pair<int, int> nl_pair = std::make_pair(n, l);
+                nl_pairs.push_back(nl_pair);
+            }
+        }
+    }
+    return nl_pairs;
+}
+
+void HydrogenRadials::smooth(std::vector<double>& rgrid,
```

Comment:
smooth of orbital is also a important part for two center integration, because any discontinunities will result in large error in FFT-based spherical Bessel transform